### PR TITLE
Encoding pretty printer

### LINKED ISSUE #35: Write a cool pretty printer for an `Encoding`
One of the nicest things about this library is that the `Encoding` type are really just "deep embeddings", or as we like to call them: syntax trees. This allows a variety of 'fun' things.

When you get an `Encoding`, that's really something like a function `Tokens -> Tokens`, and you apply a `TkEnd` to get a `Tokens` you can traverse over recursively.

It would be neat if we had a way to 'pretty print' this `Encoding` into a representation of what will actually be a CBOR value. This would be quite useful for visualising how a particular data type would be encoded into CBOR, or merely how it's structurally represented internally.

### PR DESCRIPTION
**Pull request by axman6**
Fixes #35

### CODE REVIEW

File: Data/Binary/Serialise/CBOR/Write.hs [hunk 1]
```diff
@@ -587,3 +593,210 @@ integerToBytes n0
...
+  Left s -> Left s
+  Right (TkEnd,_,ss,_) -> Right (ss "")
+  Right (toks,_,_,_) -> Left $ "prettyEnc: Not all input was consumed (this is probably a problem with the pretty printing code). Tokens left: " ++ show toks
+
+runPP :: PP a -> Encoding -> Either String (Tokens, Int, String -> String, a)
+runPP (PP f) (Encoding enc) = f (enc TkEnd) 0 id
+
+deriving instance Functor PP
+
+instance Applicative PP where
```
Comment (dcoutts) [inline Data/Binary/Serialise/CBOR/Write.hs]: Need to import Control.Applicative for older ghc/base versions.

**Comment by dcoutts** [comment]
```
case prettyEnc $ encode (3,[4]) of Right x -> putStrLn x
```
Very nice
```
82   # list(2)
   03   # integer(3)
   9f   # list(*)
      04   # integer(4)
   ff   # break
```

File: Data/Binary/Serialise/CBOR/Write.hs (hunk 1 above)
Comment (axman6) [inline Data/Binary/Serialise/CBOR/Write.hs]: Fixed

**Comment by axman6** [comment]
A slightly larger example:

```haskell
either Prelude.putStrLn Prelude.putStrLn $ prettyEnc (encode (True,[1,2,3::Int], (MP.fromList [("Hello",True),("World",False)],5::Int)))
```
```
83   # list(3)
   f5   # bool(true)
   9f   # list(*)
      01   # int(1)
      02   # int(2)
      03   # int(3)
   ff   # break
   82   # list(2)
      a2   # map(2)
         65 48 65 6c 6c 6f   # text("Hello")
         f5   # bool(true)
         65 57 6f 72 6c 64   # text("World")
         f4   # bool(false)
      05   # int(5)
```

I would like to add something that a) caps the width of long strings of hex (the Text and ByteString term suffer here) b) possibly move the comment to the next line for long binary/string data, and c) cap the length of text printed in the text term comments (though this might make it less useful)

**Comment by axman6** [comment]
The above commit caps the width of the hex portion to 16 octets of data, results look like:

```haskell
either Prelude.putStrLn Prelude.putStrLn . prettyEnc . encode $
    (True,[1,2,3::Int]
    , (MP.fromList [("Hello",True),("World",False)]
    ,"This is a very long sentense which should wrap")
    )
```
```
83   # list(3)
   f5   # bool(true)
   9f   # list(*)
      01   # int(1)
      02   # int(2)
      03   # int(3)
   ff   # break
   82   # list(2)
      a2   # map(2)
         65 48 65 6c 6c 6f   # text("Hello")
         f5   # bool(true)
         65 57 6f 72 6c 64   # text("World")
         f4   # bool(false)
      78 2e 54 68 69 73 20 69 73 20 61 20 76 65 72 79 
      20 6c 6f 6e 67 20 73 65 6e 74 65 6e 73 65 20 77 
      68 69 63 68 20 73 68 6f 75 6c 64 20 77 72 61 70   # text("This is a very long sentense which should wrap")
```

**Comment by dcoutts** [comment]
Nice.

In terms of API, I wonder if it'd be better to start from a `FlatTerm`. Thoughts @thoughtpolic?

Also, perhaps returning an Either isn't what we want here and we should instead report errors inline. Afterall if we're starting from a FlatTerm/Encoding rather than a known-valid Term then part of the utility of this as a debugging aid is to help with broken encodings. Otherwise we could just go from Encoding to Term and render from the known-valid term (which would likely simplify the implementation). But going straight from the encoding and discovering errors at some point gives us a better opportunity to report the location and nature of the problem.

**Comment by dcoutts** [comment]
Another random thought here. We have a reference implementation in the `tests/` dir that is pretty slow but pays very close attention to matching the structure of the CBOR spec. In particular it includes a representation of the CBOR tokens. That might be a useful intermediate form for the rendering here, since it gives the encoding in bytes rather directly.

**Comment by axman6** [comment]
Re: using `FlatTerm`, the reason I used `Encoding` (after several attempts to use FlatTerm) was that there wasn't a Serialise instance for `FlatTerm`, which meant I couldn't easily get the hex for the term exactly as the library encoded it. Since there is an instance of `Serialise` for `Tokens`, it's trivial to use the same machinery used to produce the CBOR `ByteString` to get the hex rep.

**Comment by dcoutts** [comment]
@axman6 I see. Makes sense.

I'm just wondering how we want to present this particular bit of the API, ie what we tell people it's for. Is it for exploring and understanding the CBOR encoding. Is it for debugging broken encodings?

**Comment by axman6** [comment]
That's a good point. I'm not sure how to handle broken streams nicely without writing a reimplementing a full decoder. I'm having a look at the Decoding and Read modules to see if it's possible to write a Decoder for the pretty printer, but that loses access to the byte representation we get for free from using Encoding. Currently this is still useful for debugging `encode` implementations, which are relatively easy to get wrong when written by hand.

**Comment by dcoutts** [comment]
@axman6 so @thoughtpolice and I were talking about this. We think it's probably best to present this as an aid for understanding the encoding, rather than for debugging broken binary files. In practice a broken binary file is going to be broken at some random but probably huge offset, and so this kind of rendering isn't going to help with that (it'd be huge). So yes, we should think about this as being to help people get an understanding of the low level encoding (which is what it does on cbor.me too of course). Then we should perhaps use the CBOR diagnostic notation for the Show instance for Encoding.

**Comment by thoughtpolice** [comment]
Right, so my current plan is to rip the diagnostic stuff we have now out of the test suite and fix it up. Then I'll expose it out of this library, and that will give a better basis for this PR to build on. It actually looks good and I like the idea, although to move it out of `.Write` I'll have to do some refactoring - I imagine we'd want these functions in a `.Pretty` module or something if possible. The module namespace in this package could do with some improvements, anyway.

So, give me a few days to shuffle pieces around and get back to you.

**Comment by axman6** [comment]
@dcoutts Sounds like a good plan to me

@thoughtpolice I was going to start that myself, but happy to hold off and then amend this PR to move things to the correct location. By the way, I really enjoyed your Cryptol talk, nice work!

**Comment by thoughtpolice** [comment]
Merged into a new `.Pretty` module - thanks for this!

**Comment by axman6** [comment]
@thoughtpolice does this mean I don't get my name on the commit? =)

**Comment by thoughtpolice** [comment]
Yeah sorry, I sort of botched the actual commit *title*. But I always try to keep original date/authorship attribution.

As a follow up, I need to add you to the AUTHORS.txt file. Will do.

**Pull request closed without merging**